src/arr/trove/json-structs.arr

```diff
@@ -0,0 +1,102 @@
+provide *
+provide-types *
+
+import string-dict as SD
+
+data JSON:
+  | j-obj(dict :: SD.StringDict<JSON>) with:
+    native(self):
+      d = self.dict
+      ret = [SD.mutable-string-dict:]
+      for map(s from d.keys().to-list()):
+        ret.set-now(s, d.get-value(s).native())
+      end
+      ret.freeze()
+    end,
+    serialize(self):
+      d = self.dict
+      l = for map(s from d.keys().to-list()):
+        '"' + s + '": ' + d.get-value(s).serialize()
+      end
+      "{" + l.join-str(", ") + "}"
+    end
+  | j-arr(l :: List<JSON>) with:
+    native(self):
+      self.l.map(lam(x): x.native() end)
+    end,
+    serialize(self):
+      "[" + self.l.map(lam(x): x.serialize() end).join-str(", ") + "]"
+    end
+  | j-num(n :: Number) with:
+    native(self):
+      self.n
+    end,
+    serialize(self):
+      # This feels like a pretty big hack.  All I want is to convert a
+      # roughnum (or a rational) as a JavaScript number, which just
+      # means cutting off the initial "~".
```

Comment:
Yep, it is definitely lossy.  But I also didn't want to restrict the nums this was allowed to work with to be fixnums (which doesn't include the JS doubles), so I had to figure out some way of representing non-integers.  Changing rationals to roughnums at least guarantees they'll fit into a JS value if something that's not Pyret uses the resulting JSON.

I also considered removing access to the constructors so that the only only way from Pyret to JSON is through `tojson` and from JSON to Pyret via `native`, so that I could control exactly what numbers are in j-nums.  However, I think that'd be bad both for using this in a pedagogical setting and for typing (since the "native" encoding is not typeable, at least in any type system Pyret is likely to use).